src/main/java/org/checkerframework/checker/dependencyinjection/DependencyInjectionAnnotatedTypeFactory.java

```diff
@@ -186,13 +242,128 @@ private void handleToMethodInvocation(
               boundToClassNames.forEach(
                   boundToClassName -> {
                     DependencyInjectionAnnotatedTypeFactory.knownBindings.put(
-                        boundClassName, boundToClassName);
+                        boundClassName,
+                        new KnownBindingsValue.KnownBindingsValueBuilder(boundToClassName).build());
                   });
             }
           });
     }
   }
 
+  /**
+   * Invoked when the {@code MethodInvocationNode} is a call to {@code
+   * com.google.inject.binder.AnnotatedBindingBuilder.annotatedWith}
+   *
+   * @param currentNode the current node in the block
+   * @param methodAccessNode the method access node
+   * @param methodArgumentNode the argument to the{@code annotatedWith} method
+   */
+  private void handleAnnotatedWithMethod(
+      Node currentNode, MethodAccessNode methodAccessNode, Node methodArgumentNode) {
+    Node receiver = methodAccessNode.getReceiver();
+
+    CFValue value = this.getStoreBefore(currentNode).getValue(JavaExpression.fromNode(receiver));
+
+    AnnotationMirror bindAnno = null;
+    if (value != null && !value.getAnnotations().isEmpty()) {
+      for (AnnotationMirror anno : value.getAnnotations()) {
+        if (AnnotationUtils.areSameByName(
+            anno, "org.checkerframework.checker.dependencyinjection.qual.BindAnnotatedWith")) {
```

Comment:
This is surprising to me. I would expect that when handling a call to `annotatedWith`, there should not (yet) be an `@BindAnnotatedWith` annotation on the receiver (but there might be an `@Bind`): the receiver should be derived from some call to `bind()`. But, that's not what this code seems to be looking for.

Another thing I'm unsure of: is it possible to chain multiple calls to `annotatedWith`? That seems like it could be possible (for example, if two different annotations are required for some reason), but I'm not sure. If that's the case, then this code needs to handle **both** `@Bind` and `@BindAnnotatedWith` annotations.

Reply:
> To answer your first concern, handleAnnotatedWithMethod is only called when we come across a node that is a call to toInstance

This is an extremely confusing name, then: it is not actually handling an `annotatedWith` method, but in fact a `toInstance` method. Please clarify the names, and I'll take another look.

Reply:
> Based on what I'm seeing, the method annotatedWith returns a LinkedBindingBuilder which does not define a method annotatedWith, which I think means that there is no way to chain multiple calls to annotatedWith.

Great, thanks for looking that up.

Reply:
However, it actually is handling the `annotatedWith` method.

```java
  private void handleToInstanceMethodInvocation(
      Node currentNode, MethodAccessNode methodAccessNode, Node methodArgumentNode) {

    Node receiver = methodAccessNode.getReceiver();

    if (isAnnotatedWithMethod(receiver.getTree())) {
      handleAnnotatedWithMethod(currentNode, methodAccessNode, methodArgumentNode);
```
When we come across a node that's a call to `toInstance`, `handleToInstanceMethodInvocation` is called. If the receiver of the node is a call to `annotatedWith`, we then call `handleAnnotatedWithMethod`. If it isn't a call to that, we treat it as a call to `bind`.

I can definitely see the confusion and it may partially be my fault for trying to keep different pieces of logic separate from each other.

Reply:
I think the problem is that "handle" is too vague a verb. Can you choose a name that's more descriptive? E.g., for the current `handleAnnotatedWithMethod`, maybe it should be `handleAnnotatedWithMethodAsReceiver` or something like that? I'm open to suggestions here, but I think the current naming scheme is too confusing.